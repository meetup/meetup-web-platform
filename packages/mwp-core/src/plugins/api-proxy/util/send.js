import fs from 'fs';

import querystring from 'qs';
import url from 'url';
import uuid from 'uuid';

import externalRequest from 'request';
import { Observable } from 'rxjs/Observable';
import 'rxjs/add/observable/of';
import 'rxjs/add/observable/bindNodeCallback';
import 'rxjs/add/observable/defer';
import 'rxjs/add/operator/catch';
import 'rxjs/add/operator/do';
import 'rxjs/add/operator/map';
import 'rxjs/add/operator/timeout';

import { API_PROXY_PLUGIN_NAME } from '../';

export const API_META_HEADER = 'X-Meta-Request-Headers';

const MOCK_RESPONSE_OK = {
	// minimal representation of http.IncomingMessage
	statusCode: 200,
	statusMessage: 'OK',
	headers: {
		'x-meetup-request-id': 'mock request',
	},
};

/*
 * When an API response is being mocked (dev only), this response will be used
 */
function makeMockResponseOk(requestOpts) {
	return {
		...MOCK_RESPONSE_OK,
		request: {
			uri: url.parse(requestOpts.url),
			method: requestOpts.method || 'get',
		},
	};
}

/**
 * In order to receive cookies from `externalRequest` requests, this function
 * provides a cookie jar that is specific to the request.
 *
 * The `requestUrl` is used to determine whether a cookie jar is needed
 *
 * https://github.com/request/request#examples
 *
 * @param {String} requestUrl the URL that will be used in the external request
 * @return {Object} a cookie jar compatible with the npm request `jar` API
 */
export const createCookieJar = requestUrl => {
	const parsedUrl = url.parse(requestUrl);
	if (parsedUrl.pathname === '/sessions') {
		return externalRequest.jar(); // create request/url-specific cookie jar
	}
	return null;
};

/**
 * Transform each query into the arguments needed for a `request` call.
 *
 * Some request options are constant for all queries, and these are curried into
 * a function that can be called with a single query as part of the request
 * stream
 *
 * @see {@link https://www.npmjs.com/package/request}
 *
 * @param {Object} externalRequestOpts request options that will be applied to
 *   every query request
 * @param {Object} query { endpoint, params, flags }
 *   call)
 * @return {Object} externalRequestOptsQuery argument for the call to
 *   `externalRequest` for the query
 */
export const buildRequestArgs = externalRequestOpts => ({
	endpoint,
	params,
	flags,
	meta = {},
}) => {
	const dataParams = querystring.stringify(params);
	const headers = { ...externalRequestOpts.headers };
	let url = encodeURI(`/${endpoint}`);
	let body;
	const jar = createCookieJar(url);

	if (flags || meta.flags) {
		headers['X-Meetup-Request-Flags'] = (flags || meta.flags).join(',');
	}

	if (meta.metaRequestHeaders) {
		headers[API_META_HEADER] = meta.metaRequestHeaders.join(',');
	}

	if (meta.variants) {
		headers['X-Meetup-Variants'] = Object.keys(
			meta.variants
		).reduce((header, experiment) => {
			const context = meta.variants[experiment];
			const contexts = context instanceof Array ? context : [context];
			header += contexts.map(c => `${experiment}=${c}`).join(' ');
			return header;
		}, '');
	}

	switch (externalRequestOpts.method) {
		case 'patch':
		case 'post':
			if (externalRequestOpts.formData) {
				break;
			}
			body = dataParams;
			headers['content-type'] = 'application/x-www-form-urlencoded';
			break;
		case 'delete':
		case 'get':
		default:
			url += `?${dataParams}`;
			headers['content-type'] = 'application/json';
			headers['X-Meta-Photo-Host'] = 'secure';
	}

	const externalRequestOptsQuery = {
		...externalRequestOpts,
		headers,
		jar,
		url,
	};

	// only add body if defined
	if (body) {
		externalRequestOptsQuery.body = body;
	}

	return externalRequestOptsQuery;
};

export function getAuthHeaders(request) {
	const oauth_token =
		request.state.oauth_token || request.plugins.requestAuth.oauth_token; // browser-based requests // internal server requests
	if (!request.state.MEETUP_MEMBER && oauth_token) {
		return {
			authorization: `Bearer ${oauth_token}`,
		};
	}
	const cookies = { ...request.state };
	const csrf = uuid.v4();
	cookies.MEETUP_CSRF = csrf;
	cookies.MEETUP_CSRF_DEV = csrf;
	const cookie = Object.keys(cookies)
		.map(name => `${name}=${cookies[name]}`)
		.join('; ');

	return {
		cookie,
		'csrf-token': csrf,
	};
}

export function getLanguageHeader(request) {
	const requestLang = request.getLanguage();
	const headerLang = request.headers['accept-language'];
	const acceptLang =
		requestLang && headerLang
			? `${requestLang},${headerLang}`
			: requestLang || headerLang;
	return acceptLang;
}

export function parseRequestHeaders(request) {
	const externalRequestHeaders = {
		...request.headers,
		...getAuthHeaders(request),
		'accept-language': getLanguageHeader(request),
	};

	delete externalRequestHeaders['host']; // let app server set 'host'
	delete externalRequestHeaders['accept-encoding']; // let app server set 'accept'
	delete externalRequestHeaders['content-length']; // original request content-length is irrelevant
	delete externalRequestHeaders['content-type']; // the content type will be set in buildRequestArgs

	return externalRequestHeaders;
}

/*
 * In multipart form requests, the parsed payload includes string key-value
 * pairs for regular inputs, and file descriptors for file upload inputs
 * ({ filename, path, headers }). This function passes through regular input
 * values unchanged, but converts the file descriptors to readable file streams
 * that will be proxied to the REST API.
 * 
 * References to the uploaded files are stored here for cleanup later on.
 */
export function parseMultipart(payload, uploadsList) {
	return Object.keys(payload).reduce((formData, key) => {
		const value = payload[key];
		if (value.filename) {
			formData[key] = {
				value: fs.createReadStream(value.path),
				options: {
					filename: value.filename,
					contentType: value.headers['content-type'],
				},
			};
			uploadsList.push(value.path);
		} else {
			formData[key] = value;
		}
		return formData;
	}, {});
}

/*
 * Translate the incoming Hapi request into an 'opts' object that can be used
 * to call `request` for the REST API. This function essentially translates the
 * app server request properties (headers, payload) into corresponding REST API
 * request properties. The resulting value will be used as the _base_ set of
 * options for _every_ parallel REST API request made by the platform
 * corresponding to single incoming request.
 * 
 * @return {Object} externalRequestOpts
 */
export function getExternalRequestOpts(request) {
	const baseUrl = request.server.settings.app.api.root_url;
	const externalRequestOpts = {
		baseUrl,
		method: request.method,
		headers: parseRequestHeaders(request), // make a copy to be immutable
		mode: 'no-cors',
		time: true, // time the request for logging
		agentOptions: {
			rejectUnauthorized: baseUrl.indexOf('.dev') === -1,
		},
	};
	if (request.mime === 'multipart/form-data') {
		// multipart form data needs special treatment
		externalRequestOpts.formData = parseMultipart(
			request.payload,
			request.plugins[API_PROXY_PLUGIN_NAME].uploads
		);
	}
	return externalRequestOpts;
}

/**
 * Fake an API request and directly return the stringified mockResponse
 */
export const makeMockRequest = mockResponse => requestOpts =>
	Observable.of([
		makeMockResponseOk(requestOpts),
		JSON.stringify(mockResponse),
	]).do(() => console.log(`MOCKING response to ${requestOpts.url}`));

const externalRequest$ = Observable.bindNodeCallback(externalRequest);
/**
 * Make a real external API request, return response body string
 */
export const makeExternalApiRequest = request => requestOpts => {
	return externalRequest$(requestOpts)
		.do(
			// log errors
			null,
			err => {
				console.error(
					JSON.stringify({
						err: err.stack,
						message: 'REST API request error',
						request: {
							id: request.id,
						},
						context: requestOpts,
					})
				);
			}
		)
		.timeout(request.server.settings.app.api.timeout)
		.map(([response, body]) => [response, body, requestOpts.jar]);
};

const logOutgoingRequest = (request, requestOpts) => {
	const { method, headers } = requestOpts;

	const parsedUrl = url.parse(requestOpts.url);
	request.server.app.logger.info(
		{
			type: 'request',
			direction: 'out',
			info: {
				headers,
				url: parsedUrl,
				method,
			},
		},
		`Outgoing request ${requestOpts.method.toUpperCase()} ${parsedUrl.pathname}`
	);
};

/*
 * Make an API request and parse the response into the expected `response`
 * object shape
 */
export const makeSend$ = request => {
	// 1. get the queries and the shared `externalRequestOpts` from the request
	//    that will be applied to all queries
	const externalRequestOpts = getExternalRequestOpts(request);

	// 2. create a function that uses `externalRequestOpts` as a base from which
	//    to build query-specific API request options objects
	const queryToRequestOpts = buildRequestArgs(externalRequestOpts);

	return query => {
		const requestOpts = queryToRequestOpts(query);
		const request$ = query.mockResponse
			? makeMockRequest(query.mockResponse)
			: makeExternalApiRequest(request);

		return Observable.defer(() => {
			logOutgoingRequest(request, requestOpts);
			return request$(requestOpts);
		});
	};
};
