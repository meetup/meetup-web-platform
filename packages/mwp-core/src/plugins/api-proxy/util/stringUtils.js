export const coerceBool = s => {
	switch (s) {
		case 'true':
			return true;
		case 'false':
			return false;
		default:
			return s;
	}
};

/**
 * simple camel case function - not really interested in edge cases, just
 * straightforward 'this-is-hyphen' to 'thisIsHyphen'
 */
export function toCamelCase(s) {
	return s.replace(/-(\w)/g, g => g[1].toUpperCase());
}
