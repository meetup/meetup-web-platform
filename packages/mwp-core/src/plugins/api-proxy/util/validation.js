import Joi from 'joi';

const stringOrArray = Joi.alternatives().try([
	Joi.string(),
	Joi.array().items(Joi.string()),
]);

export const querySchema = Joi.object({
	endpoint: Joi.string().required(),
	ref: Joi.string().required(),
	flags: Joi.array(),
	mockResponse: Joi.object(),
	params: Joi.object(), // can be FormData
	type: Joi.string(),
	meta: Joi.object({
		method: Joi.string().valid('get', 'post', 'delete', 'patch').insensitive(),
		flags: Joi.array(),
		variants: Joi.object().pattern(/\w+/, stringOrArray),
		metaRequestHeaders: Joi.array(),
	}),
});
