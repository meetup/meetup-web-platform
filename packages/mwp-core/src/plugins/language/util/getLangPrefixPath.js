// @flow

/*
 * create a function that, when called, will determine the desired
 * language-prefixed path for the request. This can be used to redirect the
 * request to the 'correct' URL.
 * 
 * If the current request.url.pathname is correct, it will be returned
 * unmodified
 */
export default (request: HapiRequest) => (): string => {
	const { supportedLangs } = request.server.settings.app;
	const requestLanguage = request.getLanguage();
	const originalPath = request.url.pathname;
	const firstPathComponent = originalPath.split('/')[1];
	// first listed locale is default
	if (requestLanguage === supportedLangs[0]) {
		// ensure that we are serving from un-prefixed URL
		if (supportedLangs.includes(firstPathComponent)) {
			const prefixedPath = originalPath.replace(`/${firstPathComponent}`, '');
			request.log(
				['info'],
				`Unncessary lang path prefix (${firstPathComponent})`
			);
			return prefixedPath;
		}
	} else if (requestLanguage !== firstPathComponent) {
		// must correct/insert the correct lang prefix
		const cleanOriginal = originalPath.replace(
			new RegExp(`^/(${supportedLangs.join('|')})/`),
			'/'
		);
		return `/${requestLanguage}${cleanOriginal}`;
	}
	return request.url.pathname;
};
