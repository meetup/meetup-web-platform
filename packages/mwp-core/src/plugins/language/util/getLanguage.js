// @flow

import querystring from 'qs';
import Accepts from 'accepts';

type ParseRequestLang = HapiRequest => ?string | false;

const getServerSettings = (request: HapiRequest) => request.server.settings.app;

/*
 * Get cookie-specified language using MEETUP_LANGUAGE cookie value
 */
export const getCookieLang: ParseRequestLang = (request: HapiRequest) => {
	const { isProd, supportedLangs } = getServerSettings(request);
	const LANGUAGE_COOKIE = isProd ? 'MEETUP_LANGUAGE' : 'MEETUP_LANGUAGE_DEV';

	const cookie = request.state[LANGUAGE_COOKIE];
	if (!cookie) {
		return;
	}
	const { language, country } = querystring.parse(cookie);
	const cookieLang = country ? `${language}-${country}` : language;
	return supportedLangs.includes(cookieLang) && cookieLang;
};

/*
 * get the URL-specified language - check for existing language path prefixes
 */
export const getUrlLang: ParseRequestLang = (request: HapiRequest) => {
	const { supportedLangs } = getServerSettings(request);
	const urlLang = request.url.path.split('/')[1];
	return supportedLangs.includes(urlLang) && urlLang;
};

/*
 * Get browser-request language using accept-language header
 */
export const getBrowserLang: ParseRequestLang = (request: HapiRequest) => {
	const { supportedLangs } = getServerSettings(request);
	return Accepts(request).language(supportedLangs);
};

/*
 * Get the language code determined from properties of the request (cookie,
 * url, header)
 */
export default (request: HapiRequest) => () =>
	// return the first language hit in the order of preference
	getCookieLang(request) || getUrlLang(request) || getBrowserLang(request);
