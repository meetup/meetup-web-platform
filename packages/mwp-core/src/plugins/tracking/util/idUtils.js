// @flow
import uuid from 'uuid';
import { ACTIVITY_PLUGIN_NAME } from '../activity';

type UpdateTrackId = TrackOpts => (Object, ?boolean) => string;
/*
 * Initialize the trackId for member or anonymous user - the longest-living id
 * we can assign to a user. Stays in place until login or logout, when it is
 * exchanged for a new trackId
 *
 *  - If the user has a tracking cookie already set, do nothing.
 *  - Otherwise, generate a new uuid and set a tracking cookie.
 */
export const updateTrackId: UpdateTrackId = (options: TrackOpts) => (
	request: Object,
	doRefresh: ?boolean
) => {
	const { trackIdCookieName } = options;
	let trackId: string = request.state[trackIdCookieName];

	if (!trackId || doRefresh) {
		// Generate a new trackId value and store in request. Cookie will be
		// set in the plugin's onResponse handler
		trackId = uuid.v4();
		request.plugins[ACTIVITY_PLUGIN_NAME].trackId = trackId;
	}
	return trackId;
};

/*
 * This function creates a new browser session id and stores it in the request.
 * A corresponding cookie will be set in the plugin's onResponse handler in
 * order to share the session cookie across browser tabs.
 */
export const newSessionId: Object => string = request => {
	const sessionId: string = uuid.v4();
	request.plugins[ACTIVITY_PLUGIN_NAME].sessionId = sessionId;
	return sessionId;
};
