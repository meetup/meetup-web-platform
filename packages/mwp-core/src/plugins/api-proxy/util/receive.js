import querystring from 'qs';

import { API_PROXY_PLUGIN_NAME } from '../';
import { coerceBool, toCamelCase } from './stringUtils';

// match escpaed unicode characters that are treated as newline literals in JS
const ESCAPED_UNICODE_NEWLINES = /\\u2028|\\u2029/g;

/**
 * Convert the X-Meetup-Variants response header into a state-ready object
 *
 * @see {@link https://meetup.atlassian.net/wiki/display/MUP/X-Meetup-Variants}
 * @returns {Object} {
 *   [experiment]: {
 *     [context (member/chapter id)]: variantName
 *   }
 * }
 */
export const parseVariantsHeader = variantsHeader =>
	variantsHeader.split(' ').reduce((variants, keyval) => {
		const [experiment, val] = keyval.split('=');
		variants[experiment] = variants[experiment] || {};
		const [context, variant] = val.split('|');
		variants[experiment][context] = variant || null;
		return variants;
	}, {});

const X_HEADERS = ['x-total-count'];

export const parseMetaHeaders = headers => {
	const meetupHeaders = Object.keys(headers)
		.filter(h => h.startsWith('x-meetup-'))
		.reduce((meta, h) => {
			const key = toCamelCase(h.replace('x-meetup-', ''));
			meta[key] = headers[h];
			return meta;
		}, {});

	// special case handling for flags
	if (meetupHeaders.flags) {
		meetupHeaders.flags = querystring.parse(meetupHeaders.flags, {
			delimiter: ', ',
			decoder: coerceBool,
		});
	}

	// special case handling for variants
	if (meetupHeaders.variants) {
		meetupHeaders.variants = parseVariantsHeader(meetupHeaders.variants);
	}

	const xHeaders = X_HEADERS.reduce((meta, h) => {
		const key = toCamelCase(h.replace('x-', ''));
		if (h in headers) {
			meta[key] = headers[h];
		}
		return meta;
	}, {});

	const linkHeader =
		headers.link &&
		headers.link.split(',').reduce((links, link) => {
			const [urlString, relString] = link.split(';');
			const url = urlString.replace(/<|>/g, '').trim();
			var rel = relString.replace(/rel="([^"]+)"/, '$1').trim();
			links[rel] = url;
			return links;
		}, {});
	const meta = {
		...meetupHeaders,
		...xHeaders,
	};

	if (linkHeader) {
		meta.link = linkHeader;
	}
	return meta;
};

/**
 * Accept an Error and return an object that will be used in place of the
 * expected API return value
 * @param {Error} err the error to populate
 * @return {Object} { value, error? }
 */
function formatApiError(err, value = null) {
	return {
		value,
		error: err.message,
	};
}

const parseBody = body => {
	if (!body) {
		return null;
	}
	// Some newline literals will not work in JS string literals - they must be
	// converted to an escaped newline character that will work end to end ('\n')
	// treat non-success HTTP code as an error
	const safeBody = body.replace(ESCAPED_UNICODE_NEWLINES, '\\n');
	return JSON.parse(safeBody);
};

/**
 * The externalRequest callback provides a `response` object and a `body` string
 * that need to be parsed in order to determine the appropriate 'value'.
 * 
 * This function determines the { value, error } based on the status code and
 * body of the response - it will always set an 'error' value when a non-2xx
 * response is received, but it may provide additional error details that are
 * included in the 'body' of the response - body error details will usually
 * be JSON parsed into `value.errors`, but that is determined by the JSON
 * returned by the REST API.
 * 
 * @return {Object} { value, error? }
 */
export const parseApiValue = ([response, body]) => {
	if (response.statusCode === 204) {
		// NoContent response type
		return { value: null };
	}

	try {
		const value = parseBody(body);
		if (response.statusCode < 200 || response.statusCode > 399) {
			return formatApiError(new Error(response.statusMessage), value);
		}
		if (value && value.problem) {
			return formatApiError(
				new Error(`API problem: ${value.problem}: ${value.details}`),
				value
			);
		}
		return { value };
	} catch (err) {
		return formatApiError(err);
	}
};

/**
 *
 * mostly error handling - any case where the API does not satisfy the
 * "api response" formatting requirement: plain object containing the requested
 * values
 *
 * This utility is specific to the response format of the API being consumed
 * @param {Array} the callback args for npm request - [response, body], where
 * `response` is an `Http.IncomingMessage` and `body` is the body text of the
 * response.
 * @return responseObj the JSON-parsed text, possibly with error info
 */
export const makeParseApiResponse = query => ([response, body]) => {
	const { value, error } = parseApiValue([response, body]);
	const meta = {
		...parseMetaHeaders(response.headers),
		endpoint: query.endpoint,
		statusCode: response.statusCode,
	};

	return {
		value,
		error,
		meta,
	};
};
/**
 * Format apiResponse to match expected state structure
 *
 * @param {Object} apiResponse JSON-parsed api response data
 */
export const makeApiResponseToQueryResponse = query => ({
	value,
	error,
	meta,
}) => ({
	type: query.type,
	ref: query.ref,
	value,
	error,
	meta,
});

export const makeLogResponse = request => ([response, body]) => {
	const {
		elapsedTime,
		request: { id, uri: { query, pathname, href }, method },
		statusCode,
	} = response;

	const logger = request.server.app.logger;
	// production logs will automatically be JSON-parsed in Stackdriver
	const log = ((statusCode >= 400 && logger.error) ||
		(statusCode >= 300 && logger.warn) ||
		logger.info)
		.bind(logger);

	log(
		{
			type: 'response',
			direction: 'in',
			info: {
				url: href,
				query: (query || '').split('&').reduce((acc, keyval) => {
					const [key, val] = keyval.split('=');
					acc[key] = val;
					return acc;
				}, {}),
				method,
				id,
				originRequestId: request.id,
				statusCode: statusCode,
				time: elapsedTime,
				body: body.length > 256 ? `${body.substr(0, 256)}...` : body,
			},
		},
		`Incoming response ${method.toUpperCase()} ${pathname} ${response.statusCode}`
	);
};

/**
 * When a tough-cookie cookie jar is provided, forward the cookies along with
 * the overall /api response back to the client
 */
export const makeInjectResponseCookies = request => ([response, _, jar]) => {
	if (!jar) {
		return;
	}
	const requestUrl = response.toJSON().request.uri.href;
	jar.getCookies(requestUrl).forEach(cookie => {
		const cookieOptions = {
			domain: cookie.domain,
			path: cookie.path,
			isHttpOnly: cookie.httpOnly,
			isSameSite: false,
			isSecure: request.server.settings.app.isProd,
			strictHeader: false, // Can't enforce RFC 6265 cookie validation on external services
		};

		request.plugins[API_PROXY_PLUGIN_NAME].setState(
			cookie.key,
			cookie.value,
			cookieOptions
		);
	});
};

export const makeReceive = request => {
	const logResponse = makeLogResponse(request);
	const injectResponseCookies = makeInjectResponseCookies(request);
	return query => {
		const parseApiResponse = makeParseApiResponse(query);
		const apiResponseToQueryResponse = makeApiResponseToQueryResponse(query);
		return response => {
			logResponse(response); // this will leak private API response data into production logs
			injectResponseCookies(response);
			try {
				return apiResponseToQueryResponse(parseApiResponse(response));
			} catch (err) {
				return {
					...formatApiError(err),
					meta: {
						endpoint: query.endpoint,
					},
				};
			}
		};
	};
};
