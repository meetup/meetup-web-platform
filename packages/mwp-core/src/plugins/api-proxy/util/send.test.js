import fs from 'fs';
import { getServer } from '../../../util/testUtils';

import 'rxjs/add/operator/toPromise';

import {
	mockQuery,
	MOCK_RENDERPROPS,
	MOCK_RENDERPROPS_UTF8,
} from 'meetup-web-mocks/lib/app';

import {
	createCookieJar,
	makeExternalApiRequest,
	buildRequestArgs,
	getAuthHeaders,
	getExternalRequestOpts,
	getLanguageHeader,
	parseMultipart,
	API_META_HEADER,
} from './send';

jest.mock('uuid', () => ({ v4: () => 'test-uuid' }));

const MOCK_HAPI_REQUEST = {
	server: getServer(),
	method: 'get',
	headers: {},
	state: {},
	plugins: {
		'api-proxy': {
			uploads: [],
		},
		requestAuth: {},
	},
	getLanguage: () => 'en-US',
};

describe('getAuthHeaders', () => {
	it('returns authorization header if no member cookie and oauth_token', () => {
		const oauth_token = 'foo';
		const authHeaders = getAuthHeaders({
			state: { oauth_token },
			plugins: { requestAuth: {} },
		});
		expect(authHeaders.authorization.startsWith('Bearer ')).toBe(true);
		expect(authHeaders.authorization.endsWith(oauth_token)).toBe(true);
	});
	it('sets MEETUP_CSRF', () => {
		const MEETUP_MEMBER = 'foo';
		const authHeaders = getAuthHeaders({
			state: { MEETUP_MEMBER },
			plugins: { requestAuth: {} },
		});
		const cookies = authHeaders.cookie.split('; ').reduce((cookies, pair) => {
			const [name, ...value] = pair.split('=');
			return {
				...cookies,
				[name]: value.join('='),
			};
		}, {});

		expect(cookies['MEETUP_CSRF']).not.toBeUndefined();
		expect(cookies['MEETUP_CSRF_DEV']).not.toBeUndefined();
		expect(authHeaders['csrf-token']).toEqual(cookies['MEETUP_CSRF']);
	});
});

describe('getLanguageHeader', () => {
	it('returns accept-language containing request.getLanguage()', () => {
		const requestLang = 'fr-FR';
		const request = {
			headers: {},
			getLanguage: () => requestLang,
		};
		expect(getLanguageHeader(request)).toEqual(requestLang);
	});
	it('prepends parsed MEMBER_LANGUAGE cookie on existing accepts-langauge', () => {
		const headerLang = 'foo';
		const requestLang = 'fr-FR';
		const request = {
			headers: { 'accept-language': headerLang },
			getLanguage: () => requestLang,
		};
		expect(getLanguageHeader(request)).toEqual(`fr-FR,${headerLang}`);
	});
	it('returns existing accepts-langauge unmodified when no language cookie', () => {
		const headerLang = 'foo';
		const request = {
			headers: { 'accept-language': headerLang },
			getLanguage: () => {},
		};
		expect(getLanguageHeader(request)).toEqual(headerLang);
	});
});

describe('buildRequestArgs', () => {
	const testQueryResults = mockQuery(MOCK_RENDERPROPS);
	const url = 'http://example.com';
	const options = {
		url,
		headers: {
			authorization: 'Bearer testtoken',
		},
		mode: 'no-cors',
	};

	it('Converts an api config to arguments for a node-request call', () => {
		let method = 'get';
		const getArgs = buildRequestArgs({ ...options, method })(testQueryResults);
		method = 'post';
		const postArgs = buildRequestArgs({ ...options, method })(testQueryResults);
		expect(getArgs).toEqual(jasmine.any(Object));
		expect(getArgs.url).toMatch(/\?.+/); // get requests will add querystring
		expect(getArgs.hasOwnProperty('body')).toBe(false); // get requests will not have a body
		expect(postArgs.url).not.toMatch(/\?.+/); // post requests will not add querystring
		expect(postArgs.body).toEqual(jasmine.any(String)); // post requests will add body string
		// post requests will add body string
		expect(postArgs.headers['content-type']).toEqual(
			'application/x-www-form-urlencoded'
		);
	});

	it('Sets X-Meetup-Request-Flags header when query has flags', () => {
		const query = {
			endpoint: 'foo',
			type: 'bar',
			params: {
				foo: 'bar',
			},
			flags: ['asdf'],
		};
		const getArgs = buildRequestArgs({ ...options, method: 'get' })(query);
		expect(getArgs.headers['X-Meetup-Request-Flags']).not.toBeUndefined();
		const postArgs = buildRequestArgs({ ...options, method: 'post' })(query);
		expect(postArgs.headers['X-Meetup-Request-Flags']).not.toBeUndefined();
	});

	it('Sets X-Meetup-Variants header when query has flags', () => {
		const experiment = 'binge-pilot';
		const context = '1234';
		const query = {
			endpoint: 'foo',
			type: 'bar',
			params: {
				foo: 'bar',
			},
			meta: {
				variants: {
					[experiment]: context,
				},
			},
		};
		const getArgs = buildRequestArgs({ ...options, method: 'get' })(query);
		expect(getArgs.headers['X-Meetup-Variants']).toEqual(
			`${experiment}=${context}`
		);
		const postArgs = buildRequestArgs({ ...options, method: 'post' })(query);
		expect(postArgs.headers['X-Meetup-Variants']).toEqual(
			`${experiment}=${context}`
		);
	});

	it('adds api meta request header with expected value from array provided in query', () => {
		const query = {
			endpoint: 'foo',
			type: 'bar',
			meta: { metaRequestHeaders: ['foo', 'bar'] },
		};
		const requestArgs = buildRequestArgs({ ...options, method: 'get' })(query);
		const requestHeaders = Object.keys(requestArgs.headers);
		const expectedApiMetaHeader = 'foo,bar';

		expect(requestHeaders).toContain(API_META_HEADER);
		expect(requestArgs.headers[API_META_HEADER]).toBe(expectedApiMetaHeader);
	});

	const testQueryResults_utf8 = mockQuery(MOCK_RENDERPROPS_UTF8);

	it('Properly encodes the URL', () => {
		const method = 'get';
		const getArgs = buildRequestArgs({ ...options, method })(
			testQueryResults_utf8
		);
		const { pathname } = require('url').parse(getArgs.url);
		expect(/^[\x00-\xFF]*$/.test(pathname)).toBe(true); // eslint-disable-line no-control-regex
	});
});

describe('getExternalRequestOpts', () => {
	it('returns the expected object from a vanilla request', () => {
		expect(getExternalRequestOpts(MOCK_HAPI_REQUEST)).toMatchSnapshot();
	});
	it('returns the expected object from a multipart request', () => {
		// most important difference is that multipart has a 'formData' key
		expect(
			getExternalRequestOpts({
				...MOCK_HAPI_REQUEST,
				mime: 'multipart/form-data',
				payload: { foo: 'bar' },
			})
		).toMatchSnapshot();
	});
});

describe('createCookieJar', () => {
	it('returns a cookie jar for /sessions endpoint', () => {
		const jar = createCookieJar('/sessions?asdfasd');
		expect(jar).not.toBeNull();
	});
	it('returns null for non-sessions endpoint', () => {
		const jar = createCookieJar('/not-sessions?asdfasd');
		expect(jar).toBeNull();
	});
});

describe('makeExternalApiRequest', () => {
	it('calls externalRequest with requestOpts', () => {
		const requestOpts = {
			foo: 'bar',
			url: 'http://example.com',
		};
		return makeExternalApiRequest(MOCK_HAPI_REQUEST)(requestOpts)
			.toPromise()
			.then(() => require('request').mock.calls.pop()[0])
			.then(arg => expect(arg).toBe(requestOpts));
	});
	it('throws an error when the API times out', () => {
		const API_TIMEOUT = 1;
		const requestOpts = {
			foo: 'bar',
			url: 'http://example.com',
		};
		return makeExternalApiRequest({
			server: {
				settings: { app: { api: { timeout: API_TIMEOUT } } },
			},
		})(requestOpts)
			.toPromise()
			.then(
				() => expect(true).toBe(false), // should not be called
				err => expect(err).toEqual(jasmine.any(Error))
			);
	});
	it('returns the requestOpts jar at array index 2', () => {
		const requestOpts = {
			foo: 'bar',
			url: 'http://example.com',
			jar: 'fooJar',
		};
		return makeExternalApiRequest(MOCK_HAPI_REQUEST)(requestOpts)
			.toPromise()
			.then(([response, body, jar]) => expect(jar).toBe(requestOpts.jar));
	});
});

describe('parseMultipart', () => {
	it('passes through non-file data unchanged', () => {
		const payload = { foo: 'bar', baz: 1 };
		expect(parseMultipart(payload, [])).toEqual(payload);
	});
	it('converts a file descriptor to a readable stream', () => {
		// have to mock 'fs' here rather than globally b/c jest needs a clean 'fs' module to run
		const originalCRS = fs.createReadStream;
		const fakeCRS = path => ({ path });
		fs.createReadStream = jest.fn(fakeCRS);
		const payload = {
			foo: 'bar',
			myfile: {
				filename: 'baz.jpg',
				headers: { 'content-type': 'smileyface' },
				path: 'path/to/file.jpg',
			},
		};
		const expectedFile = {
			value: fakeCRS(payload.myfile.path),
			options: {
				filename: payload.myfile.filename,
				contentType: payload.myfile.headers['content-type'],
			},
		};
		const uploads = [];
		expect(parseMultipart(payload, uploads)).toEqual({
			...payload,
			myfile: expectedFile,
		});
		expect(uploads).toHaveLength(1);
		expect(uploads[0]).toEqual(payload.myfile.path);
		// restore original fs.createReadStream implementation
		fs.createReadStream = originalCRS;
	});
});
