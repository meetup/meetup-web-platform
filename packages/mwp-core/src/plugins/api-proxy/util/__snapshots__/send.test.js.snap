// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getExternalRequestOpts returns the expected object from a multipart request 1`] = `
Object {
  "agentOptions": Object {
    "rejectUnauthorized": false,
  },
  "baseUrl": "https://api.dev.meetup.com",
  "formData": Object {
    "foo": "bar",
  },
  "headers": Object {
    "accept-language": "en-US",
    "cookie": "MEETUP_CSRF=test-uuid; MEETUP_CSRF_DEV=test-uuid",
    "csrf-token": "test-uuid",
  },
  "method": "get",
  "mode": "no-cors",
  "time": true,
}
`;

exports[`getExternalRequestOpts returns the expected object from a vanilla request 1`] = `
Object {
  "agentOptions": Object {
    "rejectUnauthorized": false,
  },
  "baseUrl": "https://api.dev.meetup.com",
  "headers": Object {
    "accept-language": "en-US",
    "cookie": "MEETUP_CSRF=test-uuid; MEETUP_CSRF_DEV=test-uuid",
    "csrf-token": "test-uuid",
  },
  "method": "get",
  "mode": "no-cors",
  "time": true,
}
`;
